### PR TITLE
Feature/make ranking based on troops only

### LINKED ISSUE #5: Make ranking based on troops only
Add an enum argument to specify what to rank on. E.g.

`!membercardsranked 9GULPJ9L troops`

Would rank based on troop cards only. I.e. exclude spells/buildings. 

Might need to hardcode a lookup dict for what card is what card type. Doesn't look like royale API actually distinguishes what type each card is. But double check if that's actually the case https://developer.clashroyale.com/#/documentation

Some format ideas
```
# card_level_counts = { 13: { 'totalCount': 'spellCount': x, 'buildingCount': y, 'troopCount': z }, etc. }
# OR
# card_level_counts = { "totalCount": {13: 22, 12: 14, etc.}, "troopCount": {13: 12, 12: 10, etc.}, etc.}
```

### PR DESCRIPTION
**Pull request by josephwccheng**
closes #5

### CODE REVIEW

File: clan_members_rank.py [hunk 1]
```diff
@@ -2,9 +2,11 @@
 from multiprocessing import Pool
 
 from clash_royale_client import ClashRoyaleClient
-
+from common import schemas
 
 # NOTE: we want the highest level to come first (sort descending)
+
+
```
Comment (damoster) [inline clan_members_rank.py:9]: /nit are the double blank here between comment and function necessary?

File: clan_members_rank.py [hunk 2]
```diff
@@ -71,13 +71,14 @@ def get_clan_cards_rank(self, clan_tag):
...
                 'name': member['name'],
-                'card_level_counts': self.get_card_level_counts(member['cards'])
+                'card_level_counts': self.get_card_level_counts(member['cards'], card_type_filter)
             }
             member_cards_ranked.append(member_card_levels)
             # debugging
             # print(member_card_levels)
 
         # Rank members then return results
+        # TODO move the sorting to a function so that we can unit test just the storting on it's own lol
```
Comment (damoster) [inline clan_members_rank.py:81]: So....are you going to do the TODO in this PR?

File: common/schemas.py [hunk 3]
```diff
@@ -0,0 +1 @@
+CARD_TYPE_ID_PREFIX = {'26': 'troop', '27': 'building', '28': 'spell'}
```
Comment (damoster) [inline common/schemas.py:1]: Can you add a comment above this variable explaining what the numbers represent in case we look back at this a few months later and are like....wtf....

File: main.py [hunk 4]
```diff
@@ -67,41 +69,64 @@ async def marco(ctx):
...
+        if len(args) == 1:
+            card_type_arg = args[0].rstrip('s')
+            valid_arguments = schemas.CARD_TYPE_ID_PREFIX.values()
+            if card_type_arg not in valid_arguments:
+                message = 'First argument is not in scope, please provide the correct card filter: troops, spells, or buildings'
+                await ctx.send(embed=discord.Embed(
+                    description=message,
+                    colour=discord.Colour.red()
+                ))
+                return
```
Comment (damoster) [inline main.py:102]: Weren't you going to make this into a function so that you don't have duplicate code in `membercardsranked` and `ausclan`? Can make it return an embed. And in the caller function, only send error message and early return if invalid argument

File: main.py [hunk 5]
```diff
@@ -67,41 +69,64 @@ async def marco(ctx):
...
 
-        await ctx.send(embed=create_clan_members_ranked_embed(clan_info, clan_members_ranked))
+        await ctx.send(embed=create_clan_members_ranked_embed(clan_info, clan_members_ranked, card_type_arg))
 
-    def create_clan_members_ranked_embed(clan_info, clan_members_ranked):
+    def create_clan_members_ranked_embed(clan_info, clan_members_ranked, card_type_arg='all'):
         n = 20  # Number of players to show
         top_n = clan_members_ranked[:n]
 
+        # TODO: Add a field or update description to indicate what filter is being used, bold it using markdown e.g. **filter_type**
```
Comment (damoster) [inline main.py:121]: /nit remove redundant comment

File: tests/test_clan_members_rank.py [hunk 6]
```diff
@@ -47,12 +48,14 @@ def test_compare_card_levels(self):
 
     def test_get_card_level_counts(self):
```
Comment (damoster) [inline tests/test_clan_members_rank.py:49]: You should also add a test that just test's `get_card_level_counts` similar to what I've done but with a filter added too

File: tests/test_clan_members_rank.py [hunk 7]
```diff
@@ -76,9 +79,39 @@ def test_get_clan_cards_rank(self):
...
+            ]
+        )
+
+        # Expected sort order (in terms of card counts is player3, player1, player2)
+        clan_info, member_cards_ranked = clan_members_rank.get_clan_cards_rank(
+            clan_tag, 'troops')
+        self.assertEqual(
+            clan_info, clash_royale_client_responses.CLAN_INFO_API_RESPONSE)
+        self.assertEqual([m['tag'] for m in member_cards_ranked], [
+                         '#YV9GU2VG', '#8VUG0GQRY', '#LYJVYUUUR'])
```
Comment (damoster) [inline tests/test_clan_members_rank.py:114]: Since everything in this test is exactly the same as above except for the lines I'm referring to in this comment, can't you just put these lines in above test (with different variable names ofc)

File: tests/test_clan_members_rank.py [hunk 8]
```diff
@@ -76,9 +79,39 @@ def test_get_clan_cards_rank(self):
...
+            clash_royale_client_responses.CLAN_INFO_API_RESPONSE['memberList']
+        ).thenReturn(
+            [
+                clash_royale_client_responses.PLAYER_1_RESPONSE,
+                clash_royale_client_responses.PLAYER_2_RESPONSE,
+                clash_royale_client_responses.PLAYER_3_RESPONSE
+            ]
+        )
+
+        # Expected sort order (in terms of card counts is player3, player1, player2)
```
Comment (damoster) [inline tests/test_clan_members_rank.py:108]: Don't just blindly copypasta comments and not change them if this order is no longer the case for your added test

File: tests/test_clan_members_rank.py [hunk 9]
```diff
@@ -35,6 +35,7 @@ def test_card_count_comparator(self):
         self.assertTrue(card_count_comparator(12, 7) < 0)
         self.assertTrue(card_count_comparator(12, 12) == 0)
 
+# TODO: add cases for troop/spells etc
```
Comment (damoster) [inline tests/test_clan_members_rank.py:38]: Remove no longer needed comment

**Review by damoster — CHANGES_REQUESTED** [review]
Missing a few test edge cases and tests could be more succint (see comments). Also some loose comments lying around that aren't needed (pls delete)

File: clan_members_rank.py (hunk 1 above)
Comment (josephwccheng) [reply clan_members_rank.py:9]: this is autolint hmmmm

File: common/schemas.py (hunk 3 above)
Comment (josephwccheng) [reply common/schemas.py:1]: done

File: clan_members_rank.py (hunk 2 above)
Comment (josephwccheng) [reply clan_members_rank.py:81]: done

File: clan_members_rank.py (hunk 1 above)
Comment (josephwccheng) [reply clan_members_rank.py:9]: double line before the comment so comment is right before the function

File: main.py (hunk 5 above)
Comment (josephwccheng) [reply main.py:121]: removed

File: tests/test_clan_members_rank.py (hunk 8 above)
Comment (josephwccheng) [reply tests/test_clan_members_rank.py:108]: done. noted

File: main.py (hunk 4 above)
Comment (josephwccheng) [reply main.py:102]: done by creating a function: card_type_check

File: tests/test_clan_members_rank.py (hunk 9 above)
Comment (josephwccheng) [reply tests/test_clan_members_rank.py:38]: removed

File: clan_members_rank.py [hunk 10]
```diff
@@ -1,7 +1,9 @@
 from functools import cmp_to_key
 from multiprocessing import Pool
+import copy
```
Comment (damoster) [inline clan_members_rank.py:3]: /supernit put `import` statements before `from` statements?

File: clan_members_rank.py [hunk 11]
```diff
@@ -31,18 +33,21 @@ def __init__(self):
...
-        # TODO: split into card types, e.g.
-        # card_level_counts = { 13: { 'totalCount': 'spellCount': x, 'buildingCount': y, 'troopCount': z }, etc. }
-        # OR
-        # card_level_counts = { "totalCount": {13: 22, 12: 14, etc.}, "troopCount": {13: 12, 12: 10, etc.}, etc.}
+    def get_card_level_counts(self, player_cards, card_type_filter='all'):
+        # card_type_filter check
+        valid_arguments = schemas.CARD_TYPE_ID_PREFIX.values()
+        if card_type_filter not in valid_arguments and card_type_filter not in 'all':
+            raise ValueError(
+                "function must have valid card types: all (defualt), troop, building, or spell")
```
Comment (damoster) [inline clan_members_rank.py:41]: /supernit put single quotes or escape quotes around the types. i.e.
```
"function must have valid card types: 'all' (defualt), 'troop', 'building', or 'spell'"
```

File: clan_members_rank.py [hunk 12]
```diff
@@ -71,13 +80,14 @@ def get_clan_cards_rank(self, clan_tag):
...
-                'card_level_counts': self.get_card_level_counts(member['cards'])
+                'card_level_counts': self.get_card_level_counts(member['cards'], card_type_filter)
             }
             member_cards_ranked.append(member_card_levels)
             # debugging
             # print(member_card_levels)
 
         # Rank members then return results
-        member_cards_ranked.sort(key=cmp_to_key(compare_card_levels))
+        # TODO move the sorting to a function so that we can unit test just the storting on it's own lol
```
Comment (damoster) [inline clan_members_rank.py:90]: Lmao, if you've completed this now. Why is this TODO comment still here?

File: main.py [hunk 13]
```diff
@@ -66,42 +68,65 @@ async def marco(ctx):
...
-            clan_info, clan_members_ranked = clan_members_ranker.get_clan_cards_rank(clan_tag)
-            print("Completed fetch ranked members processing")
+        await fetch_ranked_members(ctx, clan_tag, card_type_arg)
 
-        await ctx.send(embed=create_clan_members_ranked_embed(clan_info, clan_members_ranked))
+    @ bot.command()
+    async def ausclan(ctx, *args):
+        card_type_arg = 'all'
+        if len(args) == 1:
+            card_type_arg = args[0].lower().rstrip('s')
```
Comment (damoster) [inline main.py:89]: You could do the .lower.rstip('s') in the function itself by passing it card_type_check(ctx, args[0])

File: main.py [hunk 14]
```diff
@@ -66,42 +68,65 @@ async def marco(ctx):
...
 
         embed = discord.Embed(
             description=dedent('''
                 Players ranked by number of cards they have at each level.
-                Comparison start at level 13 card count. Showing the top {} players.
+                Comparison start at level 13 card count. Showing the top **{}** players.
                 Note that for clan wars 2.0 there can only be 15 players adding cards
                 for boat defenses.
-            '''.format(n)),
+                Filter Type: **{}**
```
Comment (damoster) [inline main.py:128]: Can you change this msg to `Card Count Filter Type` so it's bit more explicit

File: tests/test_clan_members_rank.py [hunk 15]
```diff
@@ -47,16 +47,73 @@ def test_compare_card_levels(self):
...
+        expected_card_level_counts = {
+            1: 1, 2: 0, 3: 0, 4: 0, 5: 1, 6: 1, 7: 1, 8: 2, 9: 6, 10: 27, 11: 28, 12: 24, 13: 12}
+        filtered_troop_expected_card_level_counts = {
+            1: 1, 2: 0, 3: 0, 4: 0, 5: 1, 6: 0, 7: 1, 8: 1, 9: 4, 10: 17, 11: 19, 12: 18, 13: 9}
+        filtered_spell_expected_card_level_counts = {
+            1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 1, 9: 1, 10: 4, 11: 7, 12: 4, 13: 2}
+        filtered_building_expected_card_level_counts = {
+            1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 1, 7: 0, 8: 0, 9: 1, 10: 6, 11: 2, 12: 2, 13: 1}
 
+        filtered_troop_expected_card_level_counts_p1 = {1: 1, 2: 0, 3: 0, 4: 1, 5: 0,
```
Comment (damoster) [inline tests/test_clan_members_rank.py:59]: /supernit, why is this formatting so aids? hahahaha can't you make it consistent with the above or is your linter doing this? 
i.e. make it something like this
```
        expected_card_level_counts = {
            1: 1, 2: 0, 3: 0, 4: 0, 5: 1, 6: 1, 7: 1, 8: 2, 9: 6, 10: 27, 11: 28, 12: 24, 13: 12}
```
Instead of
```
        expected_card_level_counts = {1: 1, 2: 0, 3: 0, 4: 0, 5: 1, 
              6: 1, 7: 1, 8: 2, 9: 6, 10: 27, 11: 28, 12: 24, 13: 12}
```

File: tests/test_clan_members_rank.py [hunk 16]
```diff
@@ -76,9 +133,38 @@ def test_get_clan_cards_rank(self):
...
-        self.assertEqual(clan_info, clash_royale_client_responses.CLAN_INFO_API_RESPONSE)
-        self.assertEqual([m['tag'] for m in member_cards_ranked], ['#LYJVYUUUR', '#YV9GU2VG', '#8VUG0GQRY'])
+        clan_info, member_cards_ranked = clan_members_rank.get_clan_cards_rank(
+            clan_tag)
+        self.assertEqual(
+            clan_info, clash_royale_client_responses.CLAN_INFO_API_RESPONSE)
+        self.assertEqual([m['tag'] for m in member_cards_ranked], [
+                         '#LYJVYUUUR', '#YV9GU2VG', '#8VUG0GQRY'])
+
+        # TODO Check why the following test passes even though the order is not correct
```
Comment (damoster) [inline tests/test_clan_members_rank.py:143]: Have you addressed this TODO now?

**Review by damoster — CHANGES_REQUESTED** [review]
Left some more comments. Can ignore the ones marked with /nit or /supernit if you like.

File: main.py (hunk 13 above)
Comment (josephwccheng) [reply main.py:89]: as discussed, leave it there

File: tests/test_clan_members_rank.py (hunk 16 above)
Comment (josephwccheng) [reply tests/test_clan_members_rank.py:143]: yeah this is done, removing the comment

File: tests/test_clan_members_rank.py (hunk 15 above)
Comment (josephwccheng) [reply tests/test_clan_members_rank.py:59]: done

File: tests/test_clan_members_rank.py [hunk 17]
```diff
@@ -47,16 +47,74 @@ def test_compare_card_levels(self):
...
+
+        # 4. Building filter test
+        card_level_counts = clan_members_rank.get_card_level_counts(
+            player_2_cards, 'building')
+        # then
+        self.assertEqual(card_level_counts,
+                         filtered_building_expected_card_level_counts)
+
+        # 5. Invalid filter test
+        with self.assertRaises(ValueError) as contect:
```
Comment (damoster) [inline tests/test_clan_members_rank.py:97]: /nit hahaha did you mean `context` instead of `contect`? lmao

**Review by damoster — APPROVED** [review]
Looks good to me. Just one variable name typo but no biggie

**Review by damoster — APPROVED** [review]
Noice. Ready to merge
